src/utils/php_server.rs

```diff
@@ -0,0 +1,43 @@
+use crate::utils::php_binaries;
+use crate::utils::php_server_cgi::start as start_cgi;
+use crate::utils::php_server_fpm::start as start_fpm;
+use crate::utils::php_server_native::start as start_native;
+use crate::utils::stop_process;
+
+use std::thread;
+use std::time;
+
+pub(crate) fn start() {
+    let php_bin = php_binaries::current();
+
+    let mut process = if php_bin.contains("-fpm") && cfg!(not(target_family = "windows")) {
+        start_fpm(php_bin)
+    } else if php_bin.contains("-cgi") {
+        start_cgi(php_bin)
+    } else {
+        start_native(php_bin)
+    };
+
+    let sleep_time = time::Duration::from_millis(1000);
+    thread::sleep(sleep_time);
+
+    match process.try_wait() {
+        Ok(Some(status)) => panic!(format!("PHP server exited with {}", status)),
+        Ok(None) => {
+            println!("PHP server seems ready");
+        }
+        Err(e) => panic!(format!(
+            "An error occured when checking PHP server health: {:?}",
+            e
+        )),
+    }
+
+    let pid = process.id();
+
+    ctrlc::set_handler(move || {
+        println!("Stopping PHP process...");
+        // TODO: I'd like this to work but there are "borrows" issue again -_-
+        //process.kill();
```

Comment:
I'm having big troubles understanding rust's "borrow" system.

When I uncomment this line, compilation fails with this error:

```
error[E0596]: cannot borrow `process` as mutable, as it is a captured variable in a `Fn` closure
  --> src/utils/php_server.rs:40:9
   |
40 |         process.kill();
   |         ^^^^^^^ cannot borrow as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
error: could not compile `rymfony`.

To learn more, run the command again with --verbose.
```

I really don't understand how to make this work: I definitely need to kill the child process: `stop_process::stop()` does not work, it will kill php-fpm's subprocesses, but not the main fpm process, and I tend to think it's because it's "attached" to rymfony's main process.

If I could kill the child process right from here, I'm certain it would work perfectly, but... borrows & stuff :sob:

Reply:
did you try `std::process::exit(exit_code);`?

Reply:
`std::process::exit` terminates the _current_ process as explained in https://doc.rust-lang.org/std/process/fn.exit.html.

Borrowing is a powerful pattern, but it is true it takes a little time to get used to it. The official manual explains it very well, in case you didn't read it yet. Or the [nomicon](https://doc.rust-lang.org/nomicon/), I highly recommend.

Your problem here is that `process` is mutable, but `ctrlc::set_handler` expects a closure that captures its environment in a non-mutable way (it's a `Fn`, not a `FnMut`). Try the following:

```rust
// Remove the mutability for `process`.
let process = process;

ctrlc::set_handler(move || { … })
```

(not tested)
```

Reply:
Thanks for the docs about the "nomicon", it seems very advanced for my little knowledge! I re-read the "References and borrowing" docs again, I'm really starting to get it, but it seems my case goes further than "simple" borrow.

When I add the `let process = process;` statement to remove mutability, I end up having another error:

```
$ cargo build
error[E0596]: cannot borrow `process` as mutable, as it is not declared as mutable
  --> src/utils/php_server.rs:41:9
   |
36 |     let process = process;
   |         ------- help: consider changing this to be mutable: `mut process`
...
41 |         process.kill();
   |         ^^^^^^^ cannot borrow as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
```

The `process` var is already mutable (defined on line 13), and if I follow your advice by tweaking it for this:

```
let mut process = &mut process;
```

I end up with another much more complex error that's too cryptic for me:

```

error[E0596]: cannot borrow `*process` as mutable, as `Fn` closures cannot mutate their captured variables
  --> src/utils/php_server.rs:43:9
   |
43 |         process.kill();
   |         ^^^^^^^ cannot borrow as mutable

warning: variable does not need to be mutable
  --> src/utils/php_server.rs:38:9
   |
38 |     let mut process = &mut process;
   |         ----^^^^^^^
   |         |
   |         help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

error[E0597]: `process` does not live long enough
  --> src/utils/php_server.rs:38:23
   |
38 |       let mut process = &mut process;
   |                         ^^^^^^^^^^^^ borrowed value does not live long enough
39 |
40 | /     ctrlc::set_handler(move || {
41 | |         println!("Stopping PHP process...");
42 | |         // TODO: I'd like this to work but there are "borrows" issue again -_-
43 | |         process.kill();
44 | |         // stop_process::stop(pid.to_string().as_ref());
45 | |     }).expect("Error setting Ctrl-C handler");
   | |______- argument requires that `process` is borrowed for `'static`
46 |   }
   |   - `process` dropped here while still borrowed

error: aborting due to 2 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0596, E0597.
For more information about an error, try `rustc --explain E0596`.
```

I'm so much confused 🦀

Reply:
I found the solution!
This PR: https://github.com/Detegr/rust-ctrlc/pull/66

So... for now I'm using another version of `ctrlc`, and I hope the PR is going to be merged 😄 

The problem in fact comes from the closure being a `Fn` one, and if you want to be able to "inject" mutable variables (which is the case for `process`, since `.kill()` seems to mutate the `Child` instance), you need a `FnMut` closure.

Problem seems to be solved :)